File: rxui-sample-java/src/main/java/com/artemzin/rxui/sample/java/MainPresenter.java
package com.artemzin.rxui.sample.java;

import android.support.annotation.NonNull;

import com.artemzin.rxui.RxUi;
import com.artemzin.rxui.sample.java.AuthService.Failure;
import com.artemzin.rxui.sample.java.AuthService.Success;

import org.javatuples.Triplet;

import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observables.ConnectableObservable;

// Since most of you use MVP, sample app uses MVP too.
class MainPresenter {

    @NonNull
    private final AuthService authService;

    @NonNull
    private final Scheduler ioScheduler;

    // As you can see, we don't need Main Thread scheduler anymore!
    MainPresenter(@NonNull AuthService authService, @NonNull Scheduler ioScheduler) {
        this.authService = authService;
        this.ioScheduler = ioScheduler;
    }

    @NonNull
    Disposable bind(MainView view) {
        final CompositeDisposable disposable = new CompositeDisposable();

        Observable<String> login = view.login().share();
        Observable<String> password = view.password().share();

        // Boolean is valid/invalid flag.
        ConnectableObservable<Triplet<String, String, Boolean>> credentials = Observable
                .combineLatest(login, password, (l, p) -> Triplet.with(l, p, !l.isEmpty() && !p.isEmpty()))
                .publish();

        Observable<Object> signInEnable = credentials
                .filter(creds -> creds.getValue2())
                .map(enable -> new Object());

        Observable<Object> signInDisable = credentials
                .filter(creds -> !creds.getValue2())
                .map(disable -> new Object());

        // You can use static import for RxUi.bind()
        disposable.add(RxUi.bind(signInEnable, view.signInEnable()));
        disposable.add(RxUi.bind(signInDisable, view.signInDisable()));

        Observable<Object> signInResult = view
                .signInClicks()
                .withLatestFrom(credentials, (click, creds) -> creds.removeFrom2()) // Leave only login and password.
                .switchMap(loginAndPassword -> authService
                        .signIn(loginAndPassword.getValue0(), loginAndPassword.getValue1())
                        .subscribeOn(ioScheduler)) // "API request".
                .share();

        disposable.add(credentials.connect());

        Observable<Success> signInSuccess = signInResult
                .filter(it -> it instanceof Success)
                .cast(Success.class);

        Observable<Failure> signInFailure = signInResult
                .filter(it -> it instanceof Failure)
                .cast(Failure.class);

        // You can use static import for RxUi.bind()
        disposable.add(RxUi.bind(signInSuccess, view.signInSuccess()));
        disposable.add(RxUi.bind(signInFailure, view.signInFailure()));

        return disposable;
    }
}

File: rxui-sample-kotlin/src/main/kotlin/com/artemzin/rxui/sample/kotlin/MainPresenter.kt
package com.artemzin.rxui.sample.kotlin

import com.artemzin.rxui.kotlin.bind
import com.artemzin.rxui.sample.kotlin.AuthService.Response.Failure
import com.artemzin.rxui.sample.kotlin.AuthService.Response.Success
import io.reactivex.Scheduler
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

class MainPresenter(private val authService: AuthService, private val ioScheduler: Scheduler) {

    fun bind(view: MainView): Disposable {
        val disposable = CompositeDisposable()

        val login = view.login.share()
        val password = view.password.share()


        // Boolean is valid/invalid flag.
        fun Pair<String, String>.valid(): Boolean {
            val (login, password) = this
            return login.isNotEmpty() && password.isNotEmpty()
        }

        val credentials = Observables
                .combineLatest(login, password)
                .publish()

        disposable += credentials
                .filter { it.valid() }
                .map { Unit }
                .bind(view.signInEnable) // YES, that short, that simple and that readable!

        disposable += credentials
                .filter { !it.valid() }
                .map { Unit }
                .startWith(Unit) // Sign In should be disabled by default.
                .bind(view.signInDisable)

        val signInResult = view
                .signInClicks
                .withLatestFrom(credentials, { click, credentials -> credentials.first to credentials.second })
                .switchMap { authService.signIn(login = it.first, password = it.second).subscribeOn(ioScheduler) }
                .share()

        disposable += credentials.connect()

        disposable += signInResult
                .filter { it is Success }
                .map { it as Success }
                .bind(view.signInSuccess)

        disposable += signInResult
                .filter { it is Failure }
                .map { it as Failure }
                .bind(view.signInFailure)

        return disposable
    }
}
